File: Python_Ver/draw_figures/showSimuResults.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat
from utils import gaussianSmooth

# Function to plot spikes (plots nonzero values)
def plot_spikes(ax, x, signal, pos, color):
    spikes = np.where((~np.isnan(signal)) & (signal != 0))[0]
    ax.scatter(x[spikes], signal[spikes]*pos, s=30, color=color, marker='|')

def showSimuResults():
    legendPos = [0.38, 0.95, 0.53, 0.04]
    plotIndex = np.arange(1200, 2400)  
    timeIndex = plotIndex / 100.0

    # ---------------------------
    # RL results
    # ---------------------------
    data_rl = loadmat(r'Python_Ver/results/Simulations_RL_0.mat')
    testM1Truth     = data_rl['testM1Truth']
    pOutputTest      = data_rl['pOutputTest']
    spkOutPredictTest= data_rl['spkOutPredictTest']
    testActions      = data_rl['testActions'].squeeze()
    motorPerformTest = data_rl['motorPerformTest'].squeeze()
    testSucRate      = float(data_rl['testSucRate'].squeeze())

    fig_rl = plt.figure("RLPP results", facecolor='w', figsize=(8, 6))
    ax1 = plt.subplot(3, 1, 1)
    ax1.plot(timeIndex, gaussianSmooth(testM1Truth[0, plotIndex], 10), 'r', linewidth=1.5)
    ax1.plot(timeIndex, pOutputTest[0, plotIndex], color=np.array([54, 56, 131]) / 255, linewidth=1.5)
    plot_spikes(ax1, timeIndex, testM1Truth[0, plotIndex], 1.4, 'r')
    plot_spikes(ax1, timeIndex, spkOutPredictTest[0, plotIndex], 1.2, np.array([54, 56, 131]) / 255)
    ax1.set_ylabel('M1 1')
    ax1.legend(["Simulated M1 firing", "RL-generated M1 firing"], loc='upper left', 
            bbox_to_anchor=(legendPos[0], legendPos[1]), ncol=2)

    ax2 = plt.subplot(3, 1, 2, sharex=ax1)
    ax2.plot(timeIndex, gaussianSmooth(testM1Truth[1, plotIndex], 10), 'r', linewidth=1.5)
    ax2.plot(timeIndex, pOutputTest[1, plotIndex], color=np.array([54, 56, 131]) / 255, linewidth=1.5)
    plot_spikes(ax2, timeIndex, testM1Truth[1, plotIndex], 1.4, 'r')
    plot_spikes(ax2, timeIndex, spkOutPredictTest[1, plotIndex], 1.2, np.array([54, 56, 131]) / 255)
    ax2.set_ylabel('M1 2')

    ax3 = plt.subplot(3, 1, 3, sharex=ax1)
    for act, col in zip([3, 2, 1], ['r', 'y', 'b']):
        temp = testActions[plotIndex].astype(float).copy()
        temp[temp != act] = np.nan
        temp[temp == act] = 1.5
        ax3.plot(timeIndex, temp, col, linewidth=8, solid_capstyle='butt')
    ax3.text(timeIndex[0] + 0.49, 1.7, 'Correct movements')
    for act, col in zip([3, 2, 1], ['r', 'y', 'b']):
        temp = motorPerformTest[plotIndex].astype(float).copy()
        temp[testActions[plotIndex] == 0] = np.nan
        temp[temp != act] = np.nan
        temp[temp == act] = 0.5
        ax3.plot(timeIndex, temp, col, linewidth=8, solid_capstyle='butt')
    ax3.text(timeIndex[0] + 0.49, 0.7, f'Predicted movements from RL outputs (Time-bin success rate: {testSucRate:.2f})')
    ax3.set_ylim([0, 2])
    ax3.legend(["Press High", "Press Low", "rest"], loc='upper left', 
            bbox_to_anchor=(0.43, 0.31), ncol=3)
    ax3.set_xlabel('Time (sec)')
    ax3.set_ylabel('Movements')
    plt.tight_layout()

    # ---------------------------
    # SL results
    # ---------------------------
    data_sl = loadmat(r'Python_Ver/results/Simulations_Sup_0.mat')
    testM1Truth     = data_sl['testM1Truth']
    pOutputTest      = data_sl['pOutputTest']
    spkOutPredictTest= data_sl['spkOutPredictTest']
    testActions      = data_sl['testActions'].squeeze()
    motorPerformTest = data_sl['motorPerformTest'].squeeze()
    testSucRate      = float(data_sl['testSucRate'].squeeze())

    fig_sl = plt.figure("Supervised learning results", facecolor='w', figsize=(8, 6))
    ax1_sl = plt.subplot(3, 1, 1)
    ax1_sl.plot(timeIndex, gaussianSmooth(testM1Truth[0, plotIndex], 10), 'r', linewidth=1.5)
    ax1_sl.plot(timeIndex, pOutputTest[0, plotIndex], color=np.array([103, 146, 70]) / 255, linewidth=1.5)
    plot_spikes(ax1_sl, timeIndex, testM1Truth[0, plotIndex], 1.4, 'r')
    plot_spikes(ax1_sl, timeIndex, spkOutPredictTest[0, plotIndex], 1.2, np.array([103, 146, 70]) / 255)
    ax1_sl.set_ylabel('M1 1')
    ax1_sl.legend(["Simulated M1 firing", "SL-predicted M1 firing"], loc='upper left', 
                bbox_to_anchor=(legendPos[0], legendPos[1]), ncol=2)

    ax2_sl = plt.subplot(3, 1, 2, sharex=ax1_sl)
    ax2_sl.plot(timeIndex, gaussianSmooth(testM1Truth[1, plotIndex], 10), 'r', linewidth=1.5)
    ax2_sl.plot(timeIndex, pOutputTest[1, plotIndex], color=np.array([103, 146, 70]) / 255, linewidth=1.5)
    plot_spikes(ax2_sl, timeIndex, testM1Truth[1, plotIndex], 1.4, 'r')
    plot_spikes(ax2_sl, timeIndex, spkOutPredictTest[1, plotIndex], 1.2, np.array([103, 146, 70]) / 255)
    ax2_sl.set_ylabel('M1 2')

    ax3_sl = plt.subplot(3, 1, 3, sharex=ax1_sl)
    for act, col in zip([3, 2, 1], ['r', 'y', 'b']):
        temp = testActions[plotIndex].astype(float).copy()
        temp[temp != act] = np.nan
        temp[temp == act] = 1.5
        ax3_sl.plot(timeIndex, temp, col, linewidth=8, solid_capstyle='butt')
    ax3_sl.text(timeIndex[0] + 0.49, 1.7, 'Correct movements')
    for act, col in zip([3, 2, 1], ['r', 'y', 'b']):
        temp = motorPerformTest[plotIndex].astype(float).copy()
        temp[testActions[plotIndex] == 0] = np.nan
        temp[temp != act] = np.nan
        temp[temp == act] = 0.5
        ax3_sl.plot(timeIndex, temp, col, linewidth=8, solid_capstyle='butt')
    ax3_sl.text(timeIndex[0] + 0.49, 0.7, f'Predicted movements from SL outputs  (Time-bin success rate: {testSucRate:.2f})')
    ax3_sl.set_ylim([0, 2])
    ax3_sl.legend(["Press High", "Press Low", "rest"], loc='upper left', 
                bbox_to_anchor=(0.43, 0.31), ncol=3)
    ax3_sl.set_xlabel('Time (sec)')
    ax3_sl.set_ylabel('Movements')
    plt.tight_layout()
    plt.show()

File: draw_figures/showSimuResults.m
clear; close all
legendPos = [0.38 0.95 0.53 0.04];
plotIndex = 1201:2400;
timeIndex = plotIndex/100;

%%
load results\Simulations_RL_1.mat

f = figure("Name","RLPP results");
f.Color = 'w';
subplot(3,1,1)
plot(timeIndex, gaussianSmooth(testM1_truth(1, plotIndex), 10), 'r', 'LineWidth',1.5)
hold on 
plot(timeIndex, pOutputTest(1, plotIndex), 'Color', [54 56 131]/255, 'LineWidth',1.5)
plotSpikes(timeIndex, testM1_truth(1, plotIndex), 1.4, 'r')
plotSpikes(timeIndex, spkOutPredictTest(1, plotIndex), 1.2, [54 56 131]/255)
box off
ylabel('M1 1')
legend(["Simulated M1 firing", "RL-generated M1 firing"], 'Position', legendPos, 'Orientation','horizontal')
subplot(3,1,2)
plot(timeIndex, gaussianSmooth(testM1_truth(2, plotIndex), 10), 'r', 'LineWidth',1.5)
hold on 
plot(timeIndex, pOutputTest(2, plotIndex), 'Color', [54 56 131]/255, 'LineWidth',1.5)
plotSpikes(timeIndex, testM1_truth(2, plotIndex), 1.4, 'r')
plotSpikes(timeIndex, spkOutPredictTest(2, plotIndex), 1.2, [54 56 131]/255)
box off
ylabel('M1 2')
subplot(3,1,3)
temp = testActions(plotIndex);
temp(temp~=3) = nan;
temp(temp==3) = 1.5;
plot(timeIndex, temp, 'r', 'LineWidth', 8);
hold on
temp = testActions(plotIndex);
temp(temp~=2) = nan;
temp(temp==2) = 1.5;
plot(timeIndex, temp, 'y', 'LineWidth', 8);
temp = testActions(plotIndex);
temp(temp~=1) = nan;
temp(temp==1) = 1.5;
plot(timeIndex, temp, 'b', 'LineWidth', 8);
text(timeIndex(1)+0.49, 1.9, 'Correct movements')

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=3) = nan;
temp(temp==3) = 0.5;
plot(timeIndex, temp, 'r', 'LineWidth', 8);

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=2) = nan;
temp(temp==2) = 0.5;
plot(timeIndex, temp, 'y', 'LineWidth', 8);

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=1) = nan;
temp(temp==1) = 0.5;
plot(timeIndex, temp, 'b', 'LineWidth', 8);
text(timeIndex(1)+0.49, 0.9, ['Predicted movements from RL outputs (Time-bin success rate: ' num2str(testSucRate, '%.2f'), ')'])

hold off
ylim([0 2])
legend(["Press High", "Press Low", "rest"], 'AutoUpdate','off', "Orientation", 'horizontal', 'Position',[0.43 0.31 0.48 0.04])
box off
xlabel('Time (sec)')
ylabel('Movements')

%%
clearvars -except legendPos plotIndex timeIndex
load results\Simulations_Sup_1.mat

f = figure("Name","Supervised learning results");
f.Color = 'w';
subplot(3,1,1)
plot(timeIndex, gaussianSmooth(testM1_truth(1, plotIndex), 10), 'r', 'LineWidth',1.5)
hold on 
plot(timeIndex, pOutputTest(1, plotIndex), 'Color', [103 146 70]/255, 'LineWidth',1.5)
plotSpikes(timeIndex, testM1_truth(1, plotIndex), 1.4, 'r')
plotSpikes(timeIndex, spkOutPredictTest(1, plotIndex), 1.2, [103 146 70]/255)
box off
ylabel('M1 1')
legend(["Simulated M1 firing", "SL-predicted M1 firing"], 'Position', legendPos, 'Orientation', 'horizontal')
subplot(3,1,2)
plot(timeIndex, gaussianSmooth(testM1_truth(2, plotIndex), 10), 'r', 'LineWidth',1.5)
hold on 
plot(timeIndex, pOutputTest(2, plotIndex), 'Color', [103 146 70]/255, 'LineWidth',1.5)
plotSpikes(timeIndex, testM1_truth(2, plotIndex), 1.4, 'r')
plotSpikes(timeIndex, spkOutPredictTest(2, plotIndex), 1.2, [103 146 70]/255)
box off
ylabel('M1 2')
subplot(3,1,3)
temp = testActions(plotIndex);
temp(temp~=3) = nan;
temp(temp==3) = 1.5;
plot(timeIndex, temp, 'r', 'LineWidth', 8);
hold on
temp = testActions(plotIndex);
temp(temp~=2) = nan;
temp(temp==2) = 1.5;
plot(timeIndex, temp, 'y', 'LineWidth', 8);
temp = testActions(plotIndex);
temp(temp~=1) = nan;
temp(temp==1) = 1.5;
plot(timeIndex, temp, 'b', 'LineWidth', 8);
text(timeIndex(1)+0.49, 1.9, 'Correct movements')

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=3) = nan;
temp(temp==3) = 0.5;
plot(timeIndex, temp, 'r', 'LineWidth', 8);

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=2) = nan;
temp(temp==2) = 0.5;
plot(timeIndex, temp, 'y', 'LineWidth', 8);

temp = motor_perform_test(plotIndex);
temp(testActions(plotIndex)==0) = nan;
temp(temp~=1) = nan;
temp(temp==1) = 0.5;
rest = plot(timeIndex, temp, 'b', 'LineWidth', 8);
text(timeIndex(1)+0.49, 0.9, ['Predicted movements from SL outputs  (Time-bin success rate: ' num2str(testSucRate, '%.2f'), ')'])

hold off
ylim([0 2])
legend(["Press High", "Press Low", "rest"], 'AutoUpdate','off', "Orientation", 'horizontal', 'Position',[0.43 0.31 0.48 0.04])
box off
xlabel('Time (sec)')
ylabel('Movements')